### DISCUSSION TITLE
Resolver a uri to ip address without DNS server?

### DISCUSSION THREAD

**Discussion by baswelsink** [cesanta/mongoose#2417] [Ask questions, get answers] [answered]
Hi,

I have a question, I want to use https and need to use an uri for this otherwise the server certificate will not be accepted. However the uri I want to use is local and not in the google dns server. I already saw an earlier discussion which suggested changing the dns server in the mg_mgr struct, however is it also possible to configure resolving to an IP without specifying the DNS server, so basically overriding the default DNS server?

**Comment by scaprile** [comment, marked as answer]
If I understand you correctly, you want the equivalent of the /etc/hosts ?
If that is the case, no, Mongoose resolves names to addresses by using a DNS server; you either provide an IP or a name + a DNS server.

**Reply by baswelsink** [reply]
Hi Scaprile, yes that's indeed correct, ok then I guess I will have to set up some DNS server. Thanks for the answer!

**Reply by scaprile** [reply]
Did you try something like this ? https://nodeployfriday.com/posts/self-signed-cert/
Mongoose does not need a DNS to resolve IP addresses, and looks like you can generate a cert with the IP as host name.

**Reply by baswelsink** [reply]
Hi Scaprile, thanks for the follow up! I did indeed see that it's possible to generate SSL certificates with IP as well but I prefer uri since it's more descriptive. Shouldn't be too much of a hassle to setup a simple DNS via docker I guess :)

**Reply by baswelsink** [reply]
Hi Scaprile, sorry to bother again. However I've tried the solution with IP based certificate. When I test it with curl it's working fine. The CA is accepted.
![image](https://github.com/cesanta/mongoose/assets/20791767/703ca34a-164f-4e23-aa46-17662b21fc2b)

This is they way I send the request. 
![image](https://github.com/cesanta/mongoose/assets/20791767/3bcfbb14-1c66-4bc7-ae60-3daf327b372b)

When I execute the app I get the following result. It mentions it's unable to verify the certificate.
![image](https://github.com/cesanta/mongoose/assets/20791767/d2302a9e-87a5-4f54-9d8a-443e1326c88e)

When I check the server logs it's mentioning my client is metioning it doesn't know the CA? 
![image](https://github.com/cesanta/mongoose/assets/20791767/8218e6ec-1016-4fed-8d2b-e444a8992e57)

It seems a bit strange to me since I'm passing the same pem file to mongoose as I'm using for curl. I've even added it to my root CAs on the machine so I can curl without passing the CA file. 
![image](https://github.com/cesanta/mongoose/assets/20791767/09bfbed3-dd40-4b09-b049-038292e0fd2a)

Do you have any idea what I might be doing wrong here? 
Btw, I'm not planning to pass username and password in the url, this is just quick testing :)

**Comment by scaprile** [comment]
Please DO NOT paste screen captures, I can't copy from them and I'm too lazy to write.
Have you been able to connect to somewhere with this code ? Why adding more unkwnowns, just use our (working) http-client example.
https://mongoose.ws/documentation/tutorials/http-client/
I see you are providing what looks like a file name, I don't know which Mongoose version you are using, so I don't know if that is correct or Mongoose has no clue on how to get to the actual cert and then is failing (I don't understand your logs, those are not Mongoose logs). If you single step, you will know, if you use a known-good client, you wouldn't be in doubt. To know why what fails is failing, check what the code does, sniff the network, study, learn. Use Wireshark, follow the handshake, etc. The SSL function code is the following:
https://github.com/openssl/openssl/blob/cd138c33d82cc889fe6a16d18806fbe939279d25/ssl/statem/statem_clnt.c#L1906-L1916
If you have an actual issue, then open an issue, follow its template and provide all info for us to reproduce it.

**Reply by baswelsink** [reply]
Hi Scaprile, 

Thanks for the quick response, I was able to connect to the server with this code by using HTTP, so was expecting adding the small TLS initialization wouldn't cause an issue. I've also checked whether I made a stupid mistake and the cert file cannot be picked up but it;s in the right location. I'm using mongoose version 7.11. Will check by just using the example code. Thanks for the quick reply!

**Reply by baswelsink** [reply]
Working now, thanks again!
